### DISCUSSION TITLE
Embed `DeploymentSpec`

### DISCUSSION THREAD

**Discussion by ionut-maxim** [kubernetes-sigs/kubebuilder#2948] [Support] [answered]
Hello all,

Is it possible to embed a `DeploymentSpec` in my own type?
E.G:
```golang
appsv1 "k8s.io/api/apps/v1"
```
```golang
// ApplicationSpec defines the desired state of Application
type ApplicationSpec struct {
	// Kubernetes Deployment Template
	Deployment appsv1.Deployment `json:"deployment,omitempty"`
}
```
I have tried this but when I try applying my own custom resource I receive this error:
`Error from server (BadRequest): error when creating "sample.yaml": Application in version "v1alpha1" cannot be handled as a Application: strict decoding error: unknown field "spec.deployment.spec.template.metadata.labels"`
EDIT:
Managed to fix this. I should've actually read the error. Removing that node from my custom resource worked

**Comment by camilamacedo86** [comment, marked as answer]
Hi @ionut-maxim ,

Embedding core Kubernetes types, like `DeploymentSpec`, inside your custom resource definition (CRD) is absolutely possible. However, when you do so, you need to be careful with how you structure the objects in your YAML files.

Your `ApplicationSpec` structure indicates that you are embedding the entire `Deployment` object and not just its spec:

```go
appsv1 "k8s.io/api/apps/v1"

type ApplicationSpec struct {
	Deployment appsv1.Deployment `json:"deployment,omitempty"`
}
```

Because of this, the structure of your custom resource should match that of a Deployment when specifying it. However, the error you encountered implies you might have added or missed some fields.

If you intended to only embed the DeploymentSpec, you might want to adjust your structure like this:

```go
// +kubebuilder:object:root=true
// +kubebuilder:subresource:status

// Application is the Schema for the applications API
type Application struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   ApplicationSpec   `json:"spec,omitempty"`
	Status ApplicationStatus `json:"status,omitempty"`
}

// ApplicationSpec defines the desired state of Application
type ApplicationSpec struct {
	DeploymentSpec appsv1.DeploymentSpec `json:"deploymentSpec,omitempty"`
}

// ApplicationStatus defines the observed state of Application
type ApplicationStatus struct {
	// Add custom status fields here
}
```

Also, please check the ideal markers for your use case [here](https://book.kubebuilder.io/reference/markers) 

Marking it as answered but if you need please feel free to re-open.
